numpy/core/src/multiarray/methods.c

```diff
@@ -63,34 +63,35 @@ NpyArg_ParseKeywords(PyObject *keys, const char *format, char **kwlist, ...)
 
 
 /*
- * Forwards an ndarray method to a the Python function
- * numpy.core._methods.<name>(...)
+ * Forwards a method call to a Python function while adding `self`:
+ * callable(self, ...)
  */
 static PyObject *
-forward_ndarray_method(PyArrayObject *self, PyObject *args, PyObject *kwds,
-                            PyObject *forwarding_callable)
+npy_forward_method(
+        PyObject *callable, PyObject *self,
+        PyObject *const *args, Py_ssize_t len_args, PyObject *kwnames)
 {
-    PyObject *sargs, *ret;
-    int i, n;
+    /* 
+     * The only thing we need to do is add `self` to args, for that allocate
+     * a new one and copy it.  `PY_VECTORCALL_ARGUMENTS_OFFSET` or a stack
```

Comment:
Yeah, I had it first, but then tried and `arr.sum()` doesn't set the flag, so I figured I leave out the code duplication.  I guess e.g. Cython might set it though.  So very happy to re-add it, also happy to do the heap optimization, but I suspect these are usually so heavy wight that it isn't noticable.

Reply:
Since the routine is called by all of the methods, it seems a nice place to actually do the optimizations...

Reply:
Was about to just add it, but now tweaked the comment.  I think what is going on is that someone promises that it is OK to change the original `self` being passed (yay).

I wonder if Python will dare changing it, I think I never even added it to any of our method calls.  It seems the flag is always lost on the way, but I am not sure its intentional.

(Not that I mind Python adding it at all, the question would just be about the stable API.)

Reply:
Oh well... added the heap version, and just made it so that it will blow up if python ever starts passing it (because that is currently true for many places anyway).

I would be happy if python starts flagging it here though, may open an issue for it (or if Python just defines that this is always valid for methods (even if never flagged).

The new commit may be a bit ugly, so happy to tweak or revert, OTOH, its just one place.